lib/libc/sys/write.2
.\"	$NetBSD$
.\"
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)write.2	8.5 (Berkeley) 4/2/94
.\"
.Dd April 3, 2010
.Dt WRITE 2
.Os
.Sh NAME
.Nm write ,
.Nm writev ,
.Nm pwrite ,
.Nm pwritev
.Nd write output
.Sh LIBRARY
.Lb libc
.Sh SYNOPSIS
.In unistd.h
.Ft ssize_t
.Fn write "int d" "const void *buf" "size_t nbytes"
.Ft ssize_t
.Fn pwrite "int d" "const void *buf" "size_t nbytes" "off_t offset"
.In sys/uio.h
.Ft ssize_t
.Fn writev "int d" "const struct iovec *iov" "int iovcnt"
.Ft ssize_t
.Fn pwritev "int d" "const struct iovec *iov" "int iovcnt" "off_t offset"
.Sh DESCRIPTION
.Fn write
attempts to write
.Fa nbytes
of data to the object referenced by the descriptor
.Fa d
from the buffer pointed to by
.Fa buf .
.Fn writev
performs the same action, but gathers the output data
from the
.Fa iovcnt
buffers specified by the members of the
.Fa iov
array: iov[0], iov[1], ..., iov[iovcnt\|-\|1].
.Fn pwrite
and
.Fn pwritev
perform the same functions, but write to the specified position in
the file without modifying the file pointer.
.Pp
For
.Fn writev
and
.Fn pwritev ,
the
.Fa iovec
structure is defined as:
.Pp
.Bd -literal -offset indent -compact
struct iovec {
	void *iov_base;
	size_t iov_len;
};
.Ed
.Pp
Each
.Fa iovec
entry specifies the base address and length of an area
in memory from which data should be written.
.Fn writev
and
.Fn pwritev
will always write a complete area before proceeding
to the next.
.Pp
On objects capable of seeking, the
.Fn write
starts at a position
given by the pointer associated with
.Fa d
(see
.Xr lseek 2 ) .
Upon return from
.Fn write ,
the pointer is incremented by the number of bytes which were written.
.Pp
Objects that are not capable of seeking always write from the current
position.
The value of the pointer associated with such an object
is undefined.
.Pp
If the real user is not the super-user, then
.Fn write
clears the set-user-id bit on a file.
This prevents penetration of system security
by a user who
.Dq captures
a writable set-user-id file
owned by the super-user.
.Pp
If
.Fn write
succeeds it will update the st_ctime and st_mtime fields of the file's
meta-data (see
.Xr stat 2 ) .
.Pp
When using non-blocking I/O on objects such as sockets that are subject
to flow control,
.Fn write
and
.Fn writev
may write fewer bytes than requested;
the return value must be noted,
and the remainder of the operation should be retried when possible.
.Sh RETURN VALUES
Upon successful completion the number of bytes which were written
is returned.
Otherwise a \-1 is returned and the global variable
.Va errno
is set to indicate the error.
.Sh ERRORS
.Fn write ,
.Fn writev ,
.Fn pwrite ,
and
.Fn pwritev
will fail and the file pointer will remain unchanged if:
.Bl -tag -width Er
.It Bq Er EAGAIN
The file was marked for non-blocking I/O,
and no data could be written immediately.
.It Bq Er EBADF
.Fa d
is not a valid descriptor open for writing.
.It Bq Er EDQUOT
The user's quota of disk blocks on the file system
containing the file has been exhausted.
.It Bq Er EFAULT
Part of
.Fa iov
or data to be written to the file
points outside the process's allocated address space.
.It Bq Er EFBIG
An attempt was made to write a file that exceeds the process's
file size limit or the maximum file size.
.It Bq Er EINTR
A signal was received before any data could be written to a slow
device.
See
.Xr sigaction 2
for more information on the interaction between signals and system
calls.
.It Bq Er EINVAL
The pointer associated with
.Fa d
was negative; or
the total length of the I/O is more than can be expressed by the ssize_t
return value.
.It Bq Er EIO
An I/O error occurred while reading from or writing to the file system.
.It Bq Er ENOSPC
There is no free space remaining on the file system
containing the file.
.It Bq Er EPIPE
An attempt is made to write to a pipe that is not open
for reading by any process; or
an attempt is made to write to a socket of type
.Dv SOCK_STREAM
that is not connected to a peer socket.
.El
.Pp
In addition,
.Fn writev
and
.Fn pwritev
may return one of the following errors:
.Bl -tag -width Er
.It Bq Er EINVAL
.Fa iovcnt
was less than or equal to 0, or greater than
.Brq Dv IOV_MAX ;
or one of the
.Fa iov_len
values in the
.Fa iov
array was negative; or
the sum of the
.Fa iov_len
values in the
.Fa iov
array overflowed a 32-bit integer.
.El
.Pp
The
.Fn pwrite
and
.Fn pwritev
calls may also return the following errors:
.Bl -tag -width Er
.It Bq Er EINVAL
The specified file offset is invalid.
.It Bq Er ESPIPE
The file descriptor is associated with a pipe, socket, or FIFO.
.El
.Sh SEE ALSO
.Xr fcntl 2 ,
.Xr lseek 2 ,
.Xr open 2 ,
.Xr pipe 2 ,
.Xr poll 2 ,
.Xr select 2 ,
.Xr sigaction 2
.Sh STANDARDS
The
.Fn write
function is expected to conform to
.St -p1003.1-88 .
The
.Fn writev
and
.Fn pwrite
functions conform to
.St -xpg4.2 .
.Sh HISTORY
The
.Fn pwritev
function call
appeared in
.Nx 1.4 .
The
.Fn pwrite
function call
appeared in
.At V.4 .
The
.Fn writev
function call
appeared in
.Bx 4.2 .
The
.Fn write
function call appeared in
.At v2 .
.Sh CAVEATS
Error checks should explicitly test for \-1.
Code such as
.Bd -literal
	while ((nr = write(fd, buf, sizeof(buf))) \*[Gt] 0)
.Ed
.Pp
is not maximally portable, as some platforms allow for
.Va nbytes
to range between
.Dv SSIZE_MAX
and
.Dv SIZE_MAX
\- 2, in which case the return value of an error-free
.Fn write
may appear as a negative number distinct from \-1.
Proper loops should use
.Bd -literal
	while ((nr = write(fd, buf, sizeof(buf))) != -1 \*[Am]\*[Am] nr != 0)
.Ed


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 6.0 to NetBSD 7.0:
	window(1): Removed from the base system in favor of tmux(1).  Use the
		pkgsrc/misc/window package from now on.  [jmmv 20120215]
	pkg_add(1): imported 20120221 version from pkgsrc [wiz 20120221]
	file(1): upgraded to 5.11 [christos 20120222]
	evbmips: Add support for Lemote Yeeloong Notebook. [nonaka 20120302]
	dbsym(8): Add option -p [bsh 20120319]
	dhcpcd(8): Import dhcpcd-5.5.5 [roy 20120323]
	ipf: upgrade to 5.1.1 with new build framework, no module and
		test support yet [christos 20120323]
	pcc: upgrade to 1.1.0.DEVEL 20120325 [plunky 20120325]
	openresolv(8): Import openresolv-3.5.1 [roy 20120327]
	dhcpcd(8): Import dhcpcd-5.5.6 [roy 20120328]
	apple: Recognize the apple disk partitioning scheme, and auto-discover
		wedges. Fix iPod disk mounting, by adding quirk.
		[christos 20120306]
	fstab: Add NAME=wedgename for fs_spec and routines to parse it. Now
		wedges can be used in /etc/fstab, even in dynamic environments
		where disks come and go and wedges get renumbered.
		[christos 20120306]
	pdisk(8): Make it work for 64 bit hosts, and make it MI.
		[christos 20120306]
	amiga: Add clockport(4) layer and a driver for A1200 on-board clockport
		a1k2cp(4). [rkujawa 20120417]
	kernel: add mpii(4), a driver for LSI Logic Fusion-MPT Message Passing
		Interface II SAS controllers.  [bouyer 20120419]
	kernel: Add support for 256 CPUs on amd64. [rmind 20120420]
	OpenSSH: Imported 6.0 [christos 20120501]
	bzip2: Imported 1.0.6 [wiz 20120507]
	amiga: Rework X-Surf driver as xsurf(4). Add support for clockport(4)
		on top of xsurf(4). [rkujawa 20120515]
	backtrace(3): Add <execinfo.h> and the backtrace family of functions
		[christos 20120526]
	kernel: Add run(4), a driver for Ralink Technology
		RT2700U/RT2800U/RT3000U USB IEEE 802.11a/b/g/n wireless
		network devices, ported from OpenBSD. [nonaka 20120530]
	bind: update to 9.9.1-P1 to fix the recent rdata exploit.
		[christos 20120604]
	empb(4): Add ELBOX Mediator PCI 1200 driver. [rkujawa 20120604]
	kernel: Add a new RLIMIT_NTHR limit to limit the number of processes
		per user and the corresponding kern.maxlwp sysctl. Expose
		kern.uidinfo.* via sysctl. [christos 20120608]
	usb(9): Added support for SMP to the USB subsytem.  By Jared D.
		McNeill and myself.  [mrg 20120609]
	kernel: {send,recv}mmsg support [christos 20120621]
	kernel: Add sysctls to avoid ipv6 DoS attacks (from OpenBSD):
		net.inet6.ip6.neighborgcthresh = 2048
		net.inet6.ip6.maxifprefixes = 16
		net.inet6.ip6.maxifdefrouters = 16
		net.inet6.ip6.maxdynroutes = 4096
		[christos 20120622]
	ata(4): added support for sata Port MultiPliers (PMP)
		[bouyer 20120702]
	kernel, libc: Add MurmurHash2 function. [rmind 20120708]
	atf(7): Import 0.16.  [jmmv 20120711]
	dhcpcd(8): Import dhcpcd-5.6.1 [roy 20120712]
	kernel: Add support for sensors to provide entropy to rnd(4)
		[pgoyette 20120715]
	tdvfb(4): Add 3Dfx Voodoo2 driver. [rkujawa 20120719]
	tdvfb(4): Add 3Dfx Voodoo Graphics (aka Voodoo1) support.
		[rkujawa 20120720]
	kernel: Change mii_statchg to take "struct ifnet *" instead of
		"device_t". [matt 20120722]
	pr(1): Add support for POSIX -f and -p options. [ginsbach 20120724]
	powerpc: Add support for FPU emulation on BookE. [matt 20120722]
	OpenSSL: Imported 1.0.1c [christos 20120726]
	kernel: safepri is dead; replaced by macro IPL_SAFEPRI. [matt 20120727]
	kernel: Add malo(4), a driver for Marvell Libertas IEEE 802.11b/g
		wireless network devices, ported from OpenBSD.
		[degroote 20120730]
	kernel, libc: Add modctl(MODCTL_EXISTS, ...) to determine if modules
		can be loaded. [jnemeth 20120803]
	modstat(8): Add -A, -a, and -e options for testing module loadability.
		[jnemeth 20120803]
	libc: Update to tzcode2012e. [christos 20120809]
	arm: Add support for BEAGLEBOARDXM (TI DM37XX cpus) boards.
		[matt 20120822]
	mfi(8): Added tagged queuing support.  [bouyer 20120823]
	mfi(8): Added support for LSI SAS2208-based controllers.
		[bouyer 20120823]
	mkdep(1): Add -P option.  [jmmv 20120826]
	share/mk: Add placeholder test programs (with some very basic test
		cases within) for common bsd.*.mk modules.  [jmmv 20120826]
	arm: Add Cortex-A9 support (a9tmr, a9wdt, armgic, armperiph).  Add
		initial MP support (can boot 2nd cpu but dies quickly).  Add
		L1 SuperSection support for ARMv6/ARMv7.  Add support for 
		BCM5301X SoC.  Add initial support for OMAP4430 and partial
		support for Freescale i.MX6.  Add new common boot/kvminit
		code.  [matt 20120831]
	dhcpcd(8): Import dhcpcd-5.6.2 [roy 20120903]
	bge(4): Add support for BCM57762, found in Apple's Thunderbolt to
		Gigabit Ethernet Adapter. PR/46961 [tsutsui 20120917]
	mfi(4): add a command pass-through ioctl, and associated COMPAT_LINUX
		support. This allows to use the LSI MegaCLI linux binary
		on a NetBSD host.  [bouyer 20120919]
	zoneinfo: Import tzdata2012f. [apb 20120927]
	kernel: add clock_nanosleep(2) [christos 20121001]
	wpa:	New wpa_supplicant(8) and hostapd(8) 1.0 [christos 20121007]

usr.bin/write/write.c
/*	$NetBSD$	*/

/*
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Jef Poskanzer and Craig Leres of the Lawrence Berkeley Laboratory.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#include <sys/cdefs.h>
#ifndef lint
__COPYRIGHT("@(#) Copyright (c) 1989, 1993\
 The Regents of the University of California.  All rights reserved.");
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)write.c	8.2 (Berkeley) 4/27/95";
#else
__RCSID("$NetBSD$");
#endif
#endif /* not lint */

#include <sys/types.h>
#include <sys/param.h>
#include <sys/stat.h>
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <signal.h>
#include <time.h>
#include <fcntl.h>
#include <paths.h>
#include <pwd.h>
#include <unistd.h>
#include <err.h>
#include <errno.h>

#include "utmpentry.h"
#include "term_chk.h"

__dead static void done(int);
static void do_write(int, const char *, const uid_t);
static void wr_fputs(char *);
static int search_utmp(char *, char *, uid_t, gid_t);
static int utmp_chk(const char *, const char *);

int
main(int argc, char **argv)
{
	time_t atime;
	uid_t myuid, uid;
	int msgsok, ttyfd;
	char *mytty;
	gid_t saved_egid = getegid();

	if (setegid(getgid()) == -1)
		err(1, "setegid");
	myuid = getuid();
	ttyfd = -1;

	mytty = check_sender(&atime, myuid, saved_egid);

	/* check args */
	switch (argc) {
	case 2:
		ttyfd = search_utmp(argv[1], mytty, myuid, saved_egid);
		break;
	case 3:
		if (!strncmp(argv[2], _PATH_DEV, strlen(_PATH_DEV)))
			argv[2] += strlen(_PATH_DEV);
		if (uid_from_user(argv[1], &uid) == -1)
			errx(1, "%s: unknown user", argv[1]);
		if (utmp_chk(argv[1], argv[2]))
			errx(1, "%s is not logged in on %s",
			    argv[1], argv[2]);
		ttyfd = term_chk(uid, argv[2], &msgsok, &atime, 0, saved_egid);
		if (ttyfd == -1)
			err(1, "%s%s", _PATH_DEV, argv[2]);
		if (myuid && !msgsok)
			errx(1, "%s has messages disabled on %s",
			    argv[1], argv[2]);
		break;
	default:
		(void)fprintf(stderr, "usage: write user [tty]\n");
		exit(1);
	}
	if (setgid(getgid()) == -1)
		err(1, "setgid");
	do_write(ttyfd, mytty, myuid);
	done(0);
	/* NOTREACHED */
#ifdef __GNUC__
	return (0);
#endif
}

/*
 * utmp_chk - checks that the given user is actually logged in on
 *     the given tty
 */
static int
utmp_chk(const char *user, const char *tty)
{
	struct utmpentry *ep;

	(void)getutentries(NULL, &ep);

	for (; ep; ep = ep->next)
		if (strcmp(user, ep->name) == 0 && strcmp(tty, ep->line) == 0)
			return(0);
	return(1);
}

/*
 * search_utmp - search utmp for the "best" terminal to write to
 *
 * Ignores terminals with messages disabled, and of the rest, returns
 * the one with the most recent access time.  Returns as value the number
 * of the user's terminals with messages enabled, or -1 if the user is
 * not logged in at all.
 *
 * Special case for writing to yourself - ignore the terminal you're
 * writing from, unless that's the only terminal with messages enabled.
 */
static int
search_utmp(char *user, char *mytty, uid_t myuid, gid_t saved_egid)
{
	char tty[MAXPATHLEN];
	time_t bestatime, atime;
	int nloggedttys, nttys, msgsok, user_is_me;
	struct utmpentry *ep;
	int fd, nfd;
	uid_t uid;

	if (uid_from_user(user, &uid) == -1)
		errx(1, "%s: unknown user", user);

	(void)getutentries(NULL, &ep);

	nloggedttys = nttys = 0;
	bestatime = 0;
	user_is_me = 0;
	fd = -1;
	for (; ep; ep = ep->next)
		if (strcmp(user, ep->name) == 0) {
			++nloggedttys;
			nfd = term_chk(uid, ep->line, &msgsok, &atime, 0,
			    saved_egid);
			if (nfd == -1)
				continue;	/* bad term? skip */
			if (myuid && !msgsok) {
				close(nfd);
				continue;	/* skip ttys with msgs off */
			}
			if (strcmp(ep->line, mytty) == 0) {
				user_is_me = 1;
				if (fd == -1)
					fd = nfd;
				else
					close(nfd);
				continue;	/* don't write to yourself */
			}
			++nttys;
			if (atime > bestatime) {
				bestatime = atime;
				(void)strlcpy(tty, ep->line, sizeof(tty));
				close(fd);
				fd = nfd;
			} else
				close(nfd);
		}

	if (nloggedttys == 0)
		errx(1, "%s is not logged in", user);
	if (nttys == 0) {
		if (user_is_me)			/* ok, so write to yourself! */
			return fd;
		errx(1, "%s has messages disabled", user);
	} else if (nttys > 1)
		warnx("%s is logged in more than once; writing to %s",
		    user, tty);
	return fd;
}

/*
 * do_write - actually make the connection
 */
static void
do_write(int ttyfd, const char *mytty, const uid_t myuid)
{
	const char *login;
	char *nows;
	struct passwd *pwd;
	time_t now;
	char host[MAXHOSTNAMELEN + 1], line[512];

	/* Determine our login name before we re-open stdout */
	if ((login = getlogin()) == NULL) {
		if ((pwd = getpwuid(myuid)) != NULL)
			login = pwd->pw_name;
		else	login = "???";
	}

	if (dup2(ttyfd, STDOUT_FILENO) == -1)
		err(1, "dup2");

	(void)signal(SIGINT, done);
	(void)signal(SIGHUP, done);
	(void)close(ttyfd);

	/* print greeting */
	if (gethostname(host, sizeof(host)) < 0)
		(void)strlcpy(host, "???", sizeof(host));
	else
		host[sizeof(host) - 1] = '\0';
	now = time(NULL);
	nows = ctime(&now);
	nows[16] = '\0';
	(void)printf("\r\n\a\a\aMessage from %s@%s on %s at %s ...\r\n",
	    login, host, mytty, nows + 11);

	while (fgets(line, sizeof(line), stdin) != NULL)
		wr_fputs(line);
}

/*
 * done - cleanup and exit
 */
static void
done(int signo)
{

	(void)write(STDOUT_FILENO, "EOF\r\n", sizeof("EOF\r\n") - 1);
	if (signo == 0)
		exit(0);
	else
		_exit(0);
}

/*
 * wr_fputs - like fputs(), but makes control characters visible and
 *     turns \n into \r\n
 */
static void
wr_fputs(char *s)
{
	unsigned char c;

#define	PUTC(c)	if (putchar(c) == EOF) goto err;

	for (; *s != '\0'; ++s) {
		c = toascii(*s);
		if (c == '\n') {
			PUTC('\r');
		} else if (!isprint(c) && !isspace(c) && c != '\a') {
			PUTC('^');
			c ^= 0x40;	/* DEL to ?, others to alpha */
		}
		PUTC(c);
	}
	return;

err:	err(1, NULL);
#undef PUTC
}
